test/HttpCacheListenerTest.php

```diff
@@ -665,4 +725,73 @@ public function testSetVary(array $cacheConfig, array $headers, array $exHeaders
 
         $this->assertSame($exHeaders, $headers->toArray());
     }
+
+
+    /**
+     * @covers \ZF\HttpCache\HttpCacheListener::setvary
+     * @dataProvider setEtagDataProvider
+     *
+     * @param array $cacheConfig
+     * @param array $headers
+     * @param array $exHeaders
+     */
+    public function testSetETag(array $cacheConfig, array $headers, array $exHeaders)
+    {
+        $this->instance->setCacheConfig($cacheConfig);
+
+        $response = new HttpResponse();
+        $headers  = $response->getHeaders()
+            ->addHeaders($headers);
+
+        $this->instance->setEtag(new HttpRequest(), $response);
+
+        $this->assertSame($exHeaders, $headers->toArray());
+    }
+
+    public function testSetETagGenerator()
+    {
+        $testGenerator = $this->prophesize(EtagGeneratorInterface::class);
+        $testGenerator->generateEtag(Argument::any(), Argument::any())->willReturn('generated');
+
+        $container = $this->prophesize(ContainerInterface::class);
+        $container->has('test-etag-generator')->willReturn(true);
+        $container->get('test-etag-generator')->willReturn($testGenerator->reveal());
+
+        $httpCacheListener = new HttpCacheListener($container->reveal());
+        $httpCacheListener->setCacheConfig([
+            'etag' => [
+                'override' => true,
+                'generator' => 'test-etag-generator'
+            ],
+        ]);
+
+        $response = new HttpResponse();
+        $headers  = $response->getHeaders();
+
+        $httpCacheListener->setEtag(new HttpRequest(), $response);
+
+        $this->assertSame(['Etag' => 'generated'], $headers->toArray());
+    }
+
+    /**
+     * @covers       \ZF\HttpCache\HttpCacheListener::setvary
+     * @dataProvider setNotModifiedDataProvider
+     *
+     * @param array $requestHeaders
+     * @param array $responseHeaders
+     * @param array $exStatusCode
+     * @internal param array $cacheConfig
+     */
+    public function testSetNotModified(array $requestHeaders, array $responseHeaders, $exStatusCode)
```

Comment:
Also write a test that demonstrates that the Not-Modified generation is skipped if the appropriate conditions are not met.